prov/util/src/util_mr_cache.c

```diff
@@ -0,0 +1,1139 @@
+/*
+ * Copyright (c) 2016-2017 Cray Inc. All rights reserved.
+ * Copyright (c) 2017 Intel Inc. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <fi_util.h>
+#include <inttypes.h>
+
+/* This macro allows to retrive a provider from the ofi_mr_cache structure */
+#define UTIL_MR_CACHE_GET_PROV(cache)	((cache)->domain->prov)
+
+#define UTIL_MR_CACHE_IS_LAZY_DEREG(cache)	((cache)->stale_size > 0)
+
+#if ENABLE_DEBUG
+#define UTIL_MR_CACHE_INC_HITS(cache)	((cache)->hits++)
+#define UTIL_MR_CACHE_INC_MISSES(cache)	((cache)->misses++)
+#else
+#define UTIL_MR_CACHE_INC_HITS(cache)	do {} while (0)
+#define UTIL_MR_CACHE_INC_MISSES(cache)	do {} while (0)
+#endif
+
+#define UTIL_MR_CACHE_DEFINE_ENTRY_FLAG_FUNCS(flag)				\
+static inline									\
+int util_mr_cache_entry_is_ ## flag (struct ofi_mr_cache_entry *entry)		\
```

Comment:
Nit: The more I see these, the less I like them. 

@shefty I recall that you might not have liked these either. I think from a readability perspective, I'd prefer they were replaced with the code they are obfuscating here.